src/main/java/com/google/sps/Scheduler.java

```diff
@@ -1,15 +1,134 @@
+    for (List<Course> validCourseList : requiredSet) {
+      generateSchedulesHelper(nonRequiredCourses, validCourseList, invariants, scheduleSet);
+    }
+
+    List<Schedule> schedules = new ArrayList<>();
+    for (List<Course> courseList: scheduleSet) {
+      schedules.add(new Schedule(courseList));
+    }
+
+    return schedules;
+  }
+
+  /**
+   * Recursively iterates through a list of courses and adds schedules to the
+   * given set.
+   * 
+   * @param courses     The list of courses to iterate through
+   * @param courseList  The current course list to be added/operated on
+   * @param invariants  The invariants that constrain the valid schedules
+   * @param courseLists The set of course lists that is being generated/added to.
+   */
+  private void generateSchedulesHelper(List<Course> courses, List<Course> courseList, 
+      Invariants invariants, Set<List<Course>> courseLists) {
+
+    // This means that we have a valid schedule, and should add it to the list.
+    float currCredits = totalCredits(courseList);
+    if (invariants.meetsCreditRequirement(currCredits)) {
+      courseLists.add(courseList);
+    }
+
+    // We've reached the end of the list or have a full schedule, we are finished.
+    if (courses.isEmpty() || currCredits >= invariants.getMaxCredits()) {
+      return;
+    }
+
+    Course course = courses.remove(0);
+    List<Section> courseSections = course.getSections();
+    List<Section> scheduleSections = courseList.stream()
+        .map(c -> c.getSections().get(0))
+        .collect(Collectors.toList());
+
+    /*
+     * Iterate over each section in the current course and check for overlap. If the
+     * section doesn't overlap and adding it would result in a valid schedule, add
+     * it and make a new recursive call.
+     */
+    // TODO: Implement logic for required courses having no valid schedules
+    for (Section section : courseSections) {
+      boolean doesNotOverlapSchedule = !sectionsOverlap(section, scheduleSections);
+      boolean doesNotExceedCreditLimit = 
+          currCredits + course.getCredits() <= invariants.getMaxCredits();
+      
+      if (doesNotOverlapSchedule && doesNotExceedCreditLimit) {
+        List<Course> newCourseList = new ArrayList<>(courseList);
+        newCourseList.add(new Course(course.getName(), course.getCourseID(), course.getSubject(),
+            course.getCredits(), course.isRequired(), Arrays.asList(section)));
+        generateSchedulesHelper(new ArrayList<>(courses), newCourseList, invariants, courseLists);
+      }
+    }
+
+    //If a course is not required, continue permuting with it *not* added to the schedule
+    if (!course.isRequired()) {
+      generateSchedulesHelper(new ArrayList<>(courses), courseList, invariants, courseLists);
+    }
+  }
+
+  /**
+   * Returns the total number of credits in a given list of courses.
+   */
+  private float totalCredits(List<Course> courseList) {
+    float total = 0;
```

Comment:
FYI - no action needed

Using `float` when you need to check for hard boundaries can cause you some problems due to the imprecise nature of floating point arithmetic. If you're curious about this, you can look into how financial systems/banks/etc handle currency calculations. The safest thing to do, if you're not going to have an arbitrarily small unit to work with, is to use ints and scale things up enough to guarantee whole numbers. In this case, if the smallest possible course credit number is 0.5 or 0.25, you could scale up all credits by a factor of 100 and use integers instead.

Here, you're *probably* going to be ok. In general this is a helpful thing to keep in mind though.